src/JustSaying/AwsTools/MessageHandling/Dispatch/MiddlewareMap.cs

```diff
@@ -0,0 +1,103 @@
+        /// Checks if a middleware has been added for a given queue and message type.
+        /// </summary>
+        /// <param name="queueName">The queue name to register the middleware for.</param>
+        /// <param name="messageType">The type of message to handle for this queue.</param>
+        /// <returns>Returns true if the middleware has been registered for the queue.</returns>
+        public bool Contains(string queueName, Type messageType)
+        {
+            if (queueName is null) throw new ArgumentNullException(nameof(queueName));
+            if (messageType is null) throw new ArgumentNullException(nameof(messageType));
+
+            return _middlewares.ContainsKey((queueName, messageType));
+        }
+
+        /// <summary>
+        /// Gets a unique list of types that are handled by all queues.
+        /// </summary>
+        public IEnumerable<Type> Types
+        {
+            get
+            {
+                var types = new HashSet<Type>();
+                foreach ((var _, Type type) in _middlewares.Keys)
+                {
+                    types.Add(type);
+                }
+
+                return types;
+            }
+        }
+
+        /// <summary>
+        /// Adds a middleware chain to be executed when a message arrives in a queue.
+        /// If the middleware is already registered for a queue, it will not be added again.
+        /// </summary>
+        /// <typeparam name="T">The type of the message to handle on this queue.</typeparam>
+        /// <param name="queueName">The queue to register the middleware for.</param>
+        /// <param name="middleware">The factory function to create middleware with.</param>
+        public MiddlewareMap Add<T>(string queueName, HandleMessageMiddleware middleware) where T : Message
+        {
+            if (queueName is null) throw new ArgumentNullException(nameof(queueName));
+            if (middleware is null) throw new ArgumentNullException(nameof(middleware));
+
+            _middlewares[(queueName, typeof(T))] = middleware;
+
+            return this;
+        }
+
+        /// <summary>
+        /// Gets a middleware factory for a queue and message type.
+        /// </summary>
+        /// <param name="queueName">The queue name to get the middleware function for.</param>
+        /// <param name="messageType">The message type to get the middleware function for.</param>
+        /// <returns>The registered middleware or null.</returns>
+        public HandleMessageMiddleware Get(string queueName, Type messageType)
+        {
+            if (queueName is null) throw new ArgumentNullException(nameof(queueName));
+            if (messageType is null) throw new ArgumentNullException(nameof(messageType));
+
+            return _middlewares.TryGetValue((queueName, messageType), out var middleware) ? middleware : null;
+        }
+
+        public InterrogationResult Interrogate()
+        {
+            var middlewares = _middlewares.Select(item =>
```

Comment:
You should `ToList()` this, otherwise it's not materialized when it's returned and could cause different results if called multiple times and/or `InvalidOperationException` to be thrown if an item is added to the dictionary mid-enumeration.

Reply:
Ah good catch, thanks 👍